---
title: "EasyMicrobiome"
author: "Zijian Wang"
date: "2/27/2023"
output: pdf_document
---
This is an in-house tutorial for microbiome data analysis based on the [microbiomeseq](https://userweb.eng.gla.ac.uk/umer.ijaz/projects/microbiomeSeq_Tutorial.html)

Current version does not contain the OTU_tree object
```{r}
# install 
#if (!"remotes" %in% installed.packages()[,1]) install.packages("remotes")
#remotes::install_github("cran/KMDA")
#remotes::install_github("cran/adespatial")

# library(devtools) # Load the devtools package
# install_github("umerijaz/microbiomeSeq") # Install the package
# library(microbiomeSeq)  #load the package
```
```{r debug}
# DEBUG
# git clone it and then revise below by using grep "ordination" -r
# delete original version: rm -rf "/Library/Frameworks/R.framework/Versions/4.1/Resources/library/microbiomeSeq"
# then install it: 
# cd "path_to_git_clone_file" #  /Users/zijianleowang/Desktop/GitHub/microbiomeSeq
# R
# install.packages(path_to_file, repos = NULL, type="source")

# restart R in RStudio: .rs.restartR()

#################### 
# edit source code
# trace(ordination, edit = T)
# from adonis_res <- vegan::adonis(dist ~ Groups, data = meta_table)
# to adonis_res <- vegan::adonis2(dist ~ Groups, data = meta_table)

# trace(plot.ordination, edit = T)
# adn_pvalue <- adn_res[[1]][["Pr(>F)"]][1]
# adn_rsquared <- round(adn_res[[1]][["R2"]][1], 3)
#adn_pvalue <- adn_res$"Pr(>F)"[1]
#adn_rsquared <- round(adn_res$"R2"[1], 3)

# trace(plot.ordination,edit=T)
# comment the lines to annotate within image
# p <- p + ggplot2::annotation_custom(grob = textGrob(label = stress.label, 
#             hjust = 0, gp = gpar(cex = 1.5, fontsize = 8)), ymin = max(ord_res$y), 
#             ymax = max(ord_res$y), xmin = extra_marginspace + 
#                 max(ord_res$x), xmax = extra_marginspace + max(ord_res$x))

# personally, I need the p-value line so, I changed
# from: out <- gridExtra::grid.arrange(p, anova_table, heights = unit.c(unit(1, 
#             "null"), th))
# to: out <- arrangeGrob(p, anova_table, heights = unit.c(unit(1, 
#             "null"), th))
# trace(generateFSO,edit=T)
# add below lines
# meta_table <- data.frame(sample_data(physeq))
#     coln.meta <- colnames(meta_table)
#     for (i in seq_along(coln.meta)){
#         meta_table[,i] <- as.numeric(meta_table[,i])
#     }


# trace(alpha_div, edit = T)
# abund_table <- as.data.frame(otu_table(physeq))

# trace(plot_cca,edit=T)
#####remove
# abund_table.adonis <- vegan::adonis(abund_table ~ ., data = meta_table)
    # bestEnvVariables <- rownames(abund_table.adonis$aov.tab)[abund_table.adonis$aov.tab$"Pr(>F)" <= 
    #     pvalueCutoff]
# bestEnvVariables <- rownames(abund_table.adonis$aov.tab)[abund_table.adonis$aov.tab$"Pr(>F)" <= 
    #     pvalueCutoff]
    # bestEnvVariables <- bestEnvVariables[!is.na(bestEnvVariables)]
#####insert
    # var = colnames(meta_table)
    # bestEnvVariables <- c()
    # for (i in seq_along(var)){
    #   abund_table.adonis <- vegan::adonis2(as.formula(paste("abund_table","~", var[i],sep=" ")), data = meta_table)
    #   if (abund_table.adonis$"Pr(>F)"[1]<=pvalueCutoff) bestEnvVariables <- c(bestEnvVariables,var[i])
    # }

####
# trace(taxa.env.correlation,edit=T)
# # taxa_list <- colnames(abund_table_filt)[1:num.taxa]
# # add one more argument as tax.name
# if (!is.null(tax.name)){
# matches <- c()
# for (element in colnames(abund_table_filt)) {
#   for (pattern in tax.name) {
#     if (grepl(pattern, element)) {
#       matches <- c(matches, element)
#     }}}
#taxa_list <- unique(c(taxa_list,matches))
}

```

```{r lib}
# load example data
library(phyloseq)
library(microbiomeSeq)
library(DESeq2)
library(vegan)
library(fso)
library(ggplot2)
library(gridExtra)
library(grid)
library(xlsx)

```

```{r self.define}
# set working dir
setwd("/Users/zijianleowang/Desktop/GitHub/EasyMicrobiome/scripts")
# set name

# pathlist = list(abundance="../data/cropps_2022_summer_tax/abundance/genus.tsv",
#                 meta="../data/cropps_2022_summer_tax/abundance/metadata.csv")
# projlist = list(projname = "maize",
#                 groupcol = "Genotype" )

pathlist = list(abundance="/Users/zijianleowang/Desktop/GitHub/NetworkAnalysis/data/Dongqi/HRSD/abundance/genus.tsv",
                meta="../data/HRSD/Metadata_HRSD.xlsx")
projlist = list(projname = "HRSD1920",
                groupcol = "VFA.to.SBPR.category" )

```

```{r func.get.data}
# load data n x d (samples x OTUs) for OTUs
get.data <- function(pathlist){
abund_table <- read.table(pathlist$abundance,sep='\t',header = TRUE,row.names = 1)
abund_table[is.na(abund_table)] <- 0

# load meta data 
pmeta = pathlist$meta
if (endsWith(pmeta, ".csv")) {
  # Read CSV file
  meta_table <- read.table(pmeta,sep=',',header=TRUE,row.names=1)
} else if (endsWith(pmeta, ".xlsx")) {
  # Read XLSX file
  meta_table <- read.xlsx(pmeta,sheetIndex = 1,row.names=1)
} else {
  # Throw an error if the file type is not supported
  stop("Unsupported file type")
}
meta_table[is.na(meta_table)] <- 0
# make them numeric
# match by intersec row name
# Original code for Maize project because it needs to be separated
# rn <- intersect(rownames(meta_table),rownames(abund_table))
# rn1.idx <- meta_table[rn,]$Project==projname ### split by project
# rn1 <- rownames(meta_table[rn,])[rn1.idx]
# meta_table <- meta_table[rn1,]
# abund_table <- abund_table[rn1,]

abund_table <- round(abund_table*1e5) # transform to counts
# check whether row name aligned
print(all(rownames(meta_table) == rownames(abund_table)))

# make it factor if not numeric
coln.meta <- colnames(meta_table)
for (i in seq_along(coln.meta)){
  if (!is.numeric(meta_table[,i]))
  meta_table[,i] <- as.factor(meta_table[,i])
}

#meta_table = as.data.frame((apply(meta_table,2,as.factor)))

# create my own physeq object
OTU = otu_table(as.matrix(abund_table),taxa_are_rows=FALSE)
SAM = sample_data(meta_table)
physeq <- phyloseq(SAM,OTU)
return (list(OTU=OTU,SAM=SAM,physeq=physeq))
}
```

```{r get data}
all = get.data(pathlist)
OTU = all$OTU
SAM = all$SAM
physeq = all$physeq
colname = colnames(sample_data(physeq)) # column names
print(colname)
```

alpha diversity of provided community data using selected indices/method(s). It performs pair-wise ANOVA of diversity measures between groups and outputs a plot for each of the selected methods(indices) annotated with significance labels.

method, options include:“richness”, “fisher”, “simpson”, “shannon” and “evenness”. It performs pairwise analysis of variance in diversity between groups and its significance annotated as on the plots. grouping_column is a categorical variable for which the grouping should be based on during the analysis. pValueCutoff specifies the p-value threshold for significance in ANOVA, default is set to 0.05. For the following examples, we use simpson, richness and shannon indices for calculating diversity.

```{r diversity}
p<-plot_anova_diversity(physeq, method = c("richness","simpson", "shannon"),
                        grouping_column =  projlist$groupcol ,pValueCutoff=0.05)
print(p)
ggsave(paste('../output/',projlist$projname,"/",'anova_div.png',sep=""),p,height=5,width=5)
```

Ordination: This is the clustering procedure of samples to detect features that are more like each other in the dataset. We implement Non-metric multidimensional Scaling (NMDS) which is a rank based approach and PCoA also known as metric/classical multidimensional scaling which uses simmilarity or dissimilarity measure to group samples and provide a representation of original dataset in a lower dimension.

Beta-dispersion: This measures variances in abundance for a group of samples by computing average distance of individual groups to the group centroid, these distances are subjected to ANOVA to test whether they are different or not.The most significantly dispersed groups are annotated on the plot with corresponding significance labels.
```{r ordination}
pvalue = 1
pg = projlist$groupcol #
method = "PCoA" #"PCoA"
ord.res <- ordination(physeq,which_distance="bray",method=method,
                      grouping_column=pg,pvalue.cutoff=pvalue)
p <- plot.ordination(ord.res, method=method, 
                     pvalue.cutoff=pvalue, show.pvalues=T,extra_marginspace=-0.5)
ggsave(paste('../output/',projlist$projname,"/",method,'Ordination_sep.png',sep=""),p,height=10,width=18)
```

FSO of environ. variables to test effects of pertubation in environmental variables to community structure. This is to show the correlation between the fuzzy set (x-axis, param, environmental variables) and original abundance (y-axis, mu, abundance) for env. variables on the top with coefficients and significance level in the bracket. The larger the coeff is, the less stronger the impacts of small changes in env. variable is on the microbial abundance. This implies that the community is very sensitive to variables with a low correlation value and
vice versa.
```{r FSO}

gp = projlist$groupcol

for (i in seq_along(colname)){
  if (!colname[i]==gp && is.numeric(as.vector(unlist(sample_data(physeq)[,i])))){
    p2 <- generateFSO(physeq, grouping_column = gp, 
                      method = 3, indices = i, filename = NULL)
    print(p2)
    ggsave(paste('../output/',projlist$projname,"/",'fso',i,'.png',sep=""),p2,height=10,width=18)
    
  }
}


```

This function finds a set of best environmental variables that describe community structure.
```{r CCA}
# trace(plot_cca,edit=T), change line 49 for font size
pcca <- plot_cca(physeq = physeq, grouping_column = projlist$groupcol, pvalueCutoff = 0.05, 
    env.variables = NULL, num.env.variables = NULL, exclude.variables = projlist$groupcol, 
    draw_species = F)
print(pcca)
ggsave(paste('../output/',projlist$projname,"/",'cca.png',sep=""),pcca,height=5,width=5)
```

```{r Corr. Env. Strain}

SAM2 = sample_data(physeq)
SAM2 = SAM2[,!(names(SAM2)) %in% c("Day","X4th.NO2.N","X4th.NO3.N")]
SAM2$artgroup = rep(1,nrow(SAM2))
physeq2 <- phyloseq(SAM2,otu_table(physeq))
env.taxa.cor <- taxa.env.correlation(physeq2, 
                                     grouping_column= projlist$groupcol,# "artgroup",
                                     method="spearman", pvalue.threshold=0.05,
                                 padjust.method="BH", adjustment=5, num.taxa=50,
                                 tax.name=c("Nitrosomonas ","Nitrosomonadaceae"),
                                 select.variables=NULL)



p <- plot_taxa_env(env.taxa.cor)
print(p)
ggsave(paste('../output/',projlist$projname,"/",'env.corr.strain.png',sep=""),p,height=10,width=10)
```